src/main/scala/zio/keeper/InternalCluster.scala

```diff
@@ -6,175 +6,292 @@ import zio.clock.Clock
+          _ => acks.delete(offset).commit
+        )
+    )
 
-          }
+  private def ack(id: Long) =
+    for {
+      promOpt <- acks
+                  .get(id)
+                  .flatMap(
+                    _.fold(STM.succeed[Option[Promise[Nothing, Unit]]](None))(prom => acks.delete(id).as(Some(prom)))
+                  )
+                  .commit
+      _ <- promOpt.fold(ZIO.unit)(_.succeed(()).unit)
     } yield ()
 
-  private def accept =
+  private def runSwim =
+    Ref.make(0).flatMap { roundRobinOffset =>
+      val loop = gossipStateRef.get.map(_.members.filterNot(_ == localMember).toIndexedSeq).flatMap { nodes =>
+        if (nodes.size > 0) {
+          for {
+            next           <- roundRobinOffset.update(old => if (old < nodes.size - 1) old + 1 else 0)
+            state          <- gossipStateRef.get
+            target         = nodes(next) // todo: insert in random position and keep going in round robin version
+            ack            <- ackMessage(10.seconds)
+            (ackId, await) = ack
+            _              <- Ping(ackId, state).serialize.flatMap(sendMessage(target.nodeId, 1, _))
+            _ <- await
+                  .timeoutFail(())(10.seconds)
+                  .foldM(
+                    _ => // attempt req messages
+                      for {
+                        jumps <- ZIO.collectAll(List.fill(3)(zio.random.nextInt(nodes.size).map(nodes(_))))
+                        pingReqs = jumps.map { jump =>
+                          for {
+                            ack            <- ackMessage(10.seconds)
+                            (ackId, await) = ack
+                            _              <- PingReq(target, ackId, state).serialize.flatMap(sendMessage(jump.nodeId, 1, _))
+                            _              <- await
+                          } yield ()
+                        }
+                        _ <- ZIO
+                              .raceAll(ZIO.never, pingReqs)
+                              .foldM(
+                                _ => removeMember(target),
+                                _ =>
+                                  putStrLn(
+                                    s"SWIM: Successful ping req to [ ${target.nodeId} ] through [ ${jumps.map(_.nodeId).mkString(", ")} ]"
+                                  )
+                              )
+                      } yield (),
+                    _ => putStrLn(s"SWIM: Successful ping to [ ${target.nodeId} ]")
+                  )
+          } yield ()
+        } else {
+          putStrLn("SWIM: No nodes to spread gossip to")
+        }
+      }
+      loop.repeat(ZSchedule.spaced(30.seconds))
+    }
+
+  private def acceptConnectionRequests =
+    ZManaged.environment[Transport].flatMap(_.bind(localMember.addr).orDie).flatMap { server =>
+      ZIO
+        .uninterruptibleMask { restore =>
+          for {
+            res <- server.accept.reserve
```

Comment:
I see the trick now. I was trying to avoid this. I don't want to merge this if we have better solution. It's adding unnecessary complexity IMHO